### PR TITLE
drivers: counter: added ctimer driver for lpcexpresso55s69

### LINKED ISSUE #22705: Implement counter driver for lpcxpresso55s69
**Is your enhancement proposal related to a problem? Please describe.**
The `lpcxpresso55s69` board currently lacks support for counter in zephyr.

**Describe the solution you'd like**
Implement a shim driver that adapts the mcux lpc RTC driver to the zephyr counter interface. The existing `counter_mcux_rtc.c` and `counter_mcux_gpt.c` shim drivers for kinetis and i.mx rt can be used for reference. Enable `tests/drivers/counter` for this board to ensure the counter driver builds and functions correctly.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by tobyfirth**
Added shim driver for the CTIMERs for the lpcexpresso55s69 board.

Fixes: #22705

Signed-off-by: Toby Firth <tobyjfirth@gmail.com>

### CODE REVIEW

**Comment by tobyfirth** [comment]
I have had to make a couple of changes to the 'test_multiple_alarms' test to enable it to pass for the ctimers driver. The test takes quite a long time to run as it takes about 45 seconds for each timer to loop around which is required for the test. There might be away to make this faster but I couldn't see a clock configuration that would make this the case.

File: drivers/counter/counter_mcux_ctimer.c [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+	 * modify the timer's config in the init function at runtime when the
+	 * frequency is known.
+	 */
+	struct mcux_lpc_ctimer_config *config = (struct mcux_lpc_ctimer_config *)dev->config;
+
+	ctimer_config_t ctimer_config;
+
+	CLOCK_AttachClk(config->clock_source);
+
+	uint32_t clk_freq = CLOCK_GetCTimerClkFreq(config->timer);
```
Comment (MaureenHelm) [inline drivers/counter/counter_mcux_ctimer.c:176]: Please move AttachClk to soc.c, and use the zephyr clock control driver (`drivers/clock_control/clock_control_mcux_syscon.c`) to get the frequency.

File: samples/drivers/counter/alarm/src/main.c [hunk 2]
```diff
@@ -31,11 +31,12 @@ struct counter_alarm_cfg alarm_cfg;
...
 #define TIMER DT_LABEL(DT_NODELABEL(timer0))
+#elif defined(CONFIG_COUNTER_MCUX_CTIMER)
+#define TIMER DT_LABEL(DT_NODELABEL(ctimer0))
 #endif
 
-static void test_counter_interrupt_fn(const struct device *counter_dev,
-				      uint8_t chan_id, uint32_t ticks,
-				      void *user_data)
+static void test_counter_interrupt_fn(const struct device *counter_dev, uint8_t chan_id,
+				      uint32_t ticks, void *user_data)
```
Comment (MaureenHelm) [inline samples/drivers/counter/alarm/src/main.c:39]: Please move all the whitespace changes to a separate commit

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 3]
```diff
@@ -86,6 +86,9 @@ static const char * const devices[] = {
 #ifdef CONFIG_COUNTER_XLNX_AXI_TIMER
 	LABELS_FOR_DT_COMPAT(xlnx_xps_timer_1_00_a)
 #endif
+#ifdef CONFIG_COUNTER_MCUX_CTIMER
+								LABELS_FOR_DT_COMPAT(nxp_lpc_ctimer)
```
Comment (MaureenHelm) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:90]: Please decrease the indentation to match the other entries.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for the contribution!

Please update `west.yml` to reference your hal_nxp PR, this will help fix the buildkite CI failure. See https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-module for details.

File: drivers/counter/counter_mcux_ctimer.c [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+#define NUM_CHANNELS 4
+
+struct mcux_lpc_ctimer_channel_data {
+	counter_alarm_callback_t alarm_callback;
+	void *alarm_user_data;
+};
+
+struct mcux_lpc_ctimer_data {
+	struct mcux_lpc_ctimer_channel_data channels[4];
```
[the remark this answers is not in the document]
Comment (tobyfirth) [reply drivers/counter/counter_mcux_ctimer.c]: Fixed.

File: drivers/counter/counter_mcux_ctimer.c [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+				     const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct mcux_lpc_ctimer_config *config = dev->config;
+	struct mcux_lpc_ctimer_data *data = dev->data;
+
+	uint32_t ticks = alarm_cfg->ticks;
+	uint32_t current = mcux_lpc_ctimer_read(config->base);
+
+	if (data->channels[chan_id].alarm_callback != NULL) {
+		return -EBUSY;
```
[the remark this answers is not in the document]
Comment (tobyfirth) [reply drivers/counter/counter_mcux_ctimer.c:84]: Added.

File: samples/drivers/counter/alarm/src/main.c (hunk 2 above)
Comment (tobyfirth) [reply samples/drivers/counter/alarm/src/main.c]: Should be removed now.

File: tests/drivers/counter/counter_basic_api/src/test_counter.c (hunk 3 above)
Comment (tobyfirth) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c]: Fixed.

File: drivers/counter/counter_mcux_ctimer.c (hunk 1 above)
Comment (tobyfirth) [reply drivers/counter/counter_mcux_ctimer.c]: Change to use the clock control API.

**Comment by tobyfirth** [comment]
I think I have updated west.yml correctly but the build still seems to be breaking.

File: drivers/counter/counter_mcux_ctimer.c [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+	uint32_t prescale;
+	void (*irq_config_func)(const struct device *dev);
+};
+
+static int mcux_lpc_ctimer_start(const struct device *dev)
+{
+	const struct mcux_lpc_ctimer_config *config = dev->config;
+	struct mcux_lpc_ctimer_data *data = dev->data;
+
+	for (uint8_t chan = 0; chan < NUM_CHANNELS; chan++) {
```
Comment (nordic-krch) [inline drivers/counter/counter_mcux_ctimer.c:42]: i wouldn't do that. It should be possible to setup alarm and then start the timer.

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 7]
```diff
@@ -546,7 +550,12 @@ void test_multiple_alarms_instance(const char *dev_name)
 	err = counter_set_channel_alarm(dev, 1, &alarm_cfg2);
 	zassert_equal(0, err, "%s: Counter set alarm failed", dev_name);
 
-	k_busy_wait(1.2*counter_ticks_to_us(dev, ticks * 2U));
+#ifdef CONFIG_COUNTER_MCUX_CTIMER
+	k_busy_wait((uint32_t)counter_ticks_to_us(dev, 0xFFFFFFFF));
```
Comment (nordic-krch) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:554]: this looks strange. If it's because ctimer does not support setting top value then i would just skip this test for that counter. You can do this by adding `set_top_value_capable()` to `multiple_channel_alarm_capable` check.

Comment (tobyfirth) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c:554]: That feeds back into one of my earlier comments, the ctimers could support this behaviour by making use of one of the channels to provide the resetting behaviour that setting the top value provides. This could be done by reducing the number of user usable channels from 4 to 3, reserving a channel for detecting the set top value and resetting or if the user sets a top value using one of the channels and producing an error if the user wants to use that channel as well as setting a top value.

Or I could just leave the driver as is and exclude the test as you suggest.

File: drivers/counter/counter_mcux_ctimer.c (hunk 6 above)
Comment (tobyfirth) [reply drivers/counter/counter_mcux_ctimer.c]: I have moved the code to the init function now.

**Comment by mmahadevan108** [comment]
Could you please help rebase and resolve conflicts. That should hopefully resolve CI failures

**Comment by tobyfirth** [comment]
Conflicts are now resolved. And I have properly updated west.yaml correctly so the builds run.

File: boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts [hunk 8]
```diff
@@ -109,3 +109,23 @@
 zephyr_udc0: &usbhs {
```
Comment (mmahadevan108) [inline boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts:109]: Should we add this change to `lpcxpresso55s69_ns.dts`

Comment (tobyfirth) [reply boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts:109]: Good point. Should the config option be also added to Kconfig.defconfig.lpc55S69_cpu1? I am guessing that the ctimers are also addressable for cpu1?

Also since the compile support for the imxrt6xx group of chips was added it should probably be added into the code. I don't actually know which specific boards are based on the imxrt6xx which is why I haven't added the configuration details for the ctimer/counter to be enabled with them.

Comment (MaureenHelm) [reply boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts:109]: Don't worry about rt6xx in the main repo yet, we will follow up with another PR. I asked you to go ahead and extend your hal_nxp PR to include rt6xx now because that change is really simple and then we won't have to update west.yml again in the follow-up PR.

File: drivers/counter/CMakeLists.txt [hunk 9]
```diff
@@ -24,3 +24,4 @@ zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_PIT            counter_mcux_pit
 zephyr_library_sources_ifdef(CONFIG_COUNTER_XLNX_AXI_TIMER      counter_xlnx_axi_timer.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_ESP32               counter_esp32.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_MICROCHIP_MCP7940N  rtc_mcp7940n.c)
+zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_CTIMER         counter_mcux_ctimer.c)
```
Comment (MaureenHelm) [inline drivers/counter/CMakeLists.txt:27]: Please move this line together with the other mcux drivers.

File: west.yml [hunk 10]
```diff
@@ -88,7 +88,7 @@ manifest:
       groups:
         - hal
     - name: hal_nxp
-      revision: ef99b119e4b41fa4610454b7b48bf9012ea7ae61
+      revision: pull/108/head
```
Comment (MaureenHelm) [inline west.yml:91]: The hal_nxp PR has now been merged, so you can update this to the new SHA.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Looks good, almost there.

File: boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts (hunk 8 above)
Comment (mmahadevan108) [reply boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts:109]: I am not sure if we should add to cpu1. For now we can add to _ns.dts file. Also could you please rebase to the latest mainline code to see if it addresses the CI build error.

File: drivers/clock_control/clock_control_mcux_syscon.c [hunk 11]
```diff
@@ -80,7 +81,24 @@ static int mcux_lpc_syscon_clock_control_get_subsys_rate(
...
+		*rate = CLOCK_GetCTimerClkFreq(2);
+		break;
+	case (MCUX_CTIMER3_CLK + MCUX_CTIMER_CLK_OFFSET):
+		*rate = CLOCK_GetCTimerClkFreq(3);
+		break;
+	case (MCUX_CTIMER4_CLK + MCUX_CTIMER_CLK_OFFSET):
+		*rate = CLOCK_GetCTimerClkFreq(4);
+		break;
 	}
+#endif
```
Comment (mmahadevan108) [inline drivers/clock_control/clock_control_mcux_syscon.c:101]: I think this `#endif` should move after line 99 so the closing brace for the switch command is available for all test cases.

Comment (tobyfirth) [reply drivers/clock_control/clock_control_mcux_syscon.c:100]: Didn't spot that. Fixed it now.

**Review by MaureenHelm — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

File: boards/arm/lpcxpresso55s69/doc/index.rst [hunk 12]
```diff
@@ -82,6 +82,8 @@ features:
 +-----------+------------+-------------------------------------+
 | USB       | on-chip    | USB device                          |
 +-----------+------------+-------------------------------------+
+| COUNTER   | on-chip    | counter                             |
++-----------+------------+-------------------------------------+
```
Comment (mmahadevan108) [inline boards/arm/lpcxpresso55s69/doc/index.rst:86]: Just noticed this, should we change `COUNTER` to `CTIMER`
